File: app/services/renderer.py
from __future__ import annotations

import base64
from pathlib import Path
from uuid import uuid4

import httpx
from playwright.async_api import Browser, async_playwright

from app.core.config import SERVER_CONFIG


_playwright = None
_browser: Browser | None = None
_active_firefly_video_records = 0


def get_browser_gpu_args(enabled: bool) -> list[str]:
    if not enabled:
        return ["--disable-gpu"]

    args = [
        "--ignore-gpu-blocklist",
        "--enable-gpu-rasterization",
        "--enable-zero-copy",
    ]

    import sys

    if sys.platform.startswith("win"):
        args.append("--use-angle=d3d11")
    elif sys.platform.startswith("linux"):
        args.extend(["--use-gl=egl", "--enable-features=VaapiVideoDecoder,VaapiVideoEncoder"])

    return args


def build_browser_launch_args(gpu_enabled: bool | None = None) -> list[str]:
    if gpu_enabled is None:
        gpu_enabled = SERVER_CONFIG.browser_gpu_enabled

    return [
        "--no-sandbox",
        "--disable-setuid-sandbox",
        "--disable-dev-shm-usage",
        *get_browser_gpu_args(gpu_enabled),
        "--no-first-run",
        "--no-zygote",
        "--disable-extensions",
        "--autoplay-policy=no-user-gesture-required",
        *SERVER_CONFIG.browser_gpu_extra_args,
    ]


async def download_image_as_data_url(url: str, max_redirects: int = 5) -> str:
    async with httpx.AsyncClient(
        timeout=15,
        follow_redirects=True,
        max_redirects=max_redirects,
        headers={
            "User-Agent": (
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"
            ),
            "Accept": "image/*,*/*",
            "Referer": str(httpx.URL(url).copy_with(path="/")),
        },
    ) as client:
        response = await client.get(url)

    if response.status_code != 200:
        raise RuntimeError(f"HTTP {response.status_code} from {url}")

    content = response.content
    if len(content) < 100:
        raise RuntimeError("Downloaded file too small - likely not an image")

    mime_type = response.headers.get("content-type", "image/jpeg").split(";")[0].strip()
    encoded = base64.b64encode(content).decode("ascii")
    return f"data:{mime_type};base64,{encoded}"


async def get_browser() -> Browser:
    global _playwright, _browser

    if _browser and _browser.is_connected():
        return _browser

    if _playwright is None:
        _playwright = await async_playwright().start()

    try:
        _browser = await _playwright.chromium.launch(
            headless=True,
            args=build_browser_launch_args(),
        )
        print(f"[Renderer] Browser GPU acceleration: {'enabled' if SERVER_CONFIG.browser_gpu_enabled else 'disabled'}")
    except Exception as err:
        if not SERVER_CONFIG.browser_gpu_enabled or not SERVER_CONFIG.browser_gpu_fallback:
            raise
        print(f"[Renderer] Browser GPU launch failed ({err}). Retrying with GPU disabled.")
        _browser = await _playwright.chromium.launch(
            headless=True,
            args=build_browser_launch_args(False),
        )
        print("[Renderer] Browser GPU acceleration: disabled after fallback")

    return _browser


async def render_firefly_video(params: dict) -> str:
    global _active_firefly_video_records

    if _active_firefly_video_records >= SERVER_CONFIG.max_concurrent:
        raise RuntimeError(
            f"Da dat gioi han {SERVER_CONFIG.max_concurrent} video dong thoi. Vui long thu lai sau."
        )

    _active_firefly_video_records += 1
    page = None
    try:
        browser = await get_browser()
        page = await browser.new_page(
            viewport={
                "width": int(params["width"]),
                "height": int(params["height"]),
                "deviceScaleFactor": 1,
            }
        )

        render_page_path = SERVER_CONFIG.public_dir / "firefly-render.html"
        await page.goto(render_page_path.resolve().as_uri(), wait_until="domcontentloaded", timeout=15000)
        await page.wait_for_function("window.__PAGE_READY === true", timeout=10000)

        firefly_config = {
            "bgIndex": params["bgIndex"],
            "bgCustom": params.get("bgUrl") or None,
            "count": params["count"],
            "size": params["size"],
            "speed": params["speed"],
            "colorMode": params["colorMode"],
            "colorIndex": params["colorIndex"],
            "customColor": params["customColor"],
            "glowLevel": params["glowLevel"],
            "direction": params["direction"],
            "spread": params["spread"],
        }

        if params.get("bgUrl"):
            print(f"[Renderer] Downloading background: {params['bgUrl']}")
            try:
                data_url = await download_image_as_data_url(params["bgUrl"])
                firefly_config["bgCustom"] = data_url
                print(f"[Renderer] Background loaded ({len(data_url) // 1024} KB data URL)")
            except Exception as err:
                print(f"[Renderer] Background download failed: {err}. Using preset instead.")
                firefly_config["bgCustom"] = None

        await page.evaluate("(cfg) => window.__applyConfig(cfg)", firefly_config)
        await page.wait_for_timeout(1500)

        print(
            "[Renderer] Starting recording: "
            f"{params['width']}x{params['height']}, {params['duration']}s, "
            f"{params['fps']}fps, {params['bitrate']}bps"
        )

        base64_data = await page.evaluate(
            """(args) => window.__startRecording(args.duration, args.fps, args.bitrate)""",
            {
                "duration": params["duration"],
                "fps": params["fps"],
                "bitrate": params["bitrate"],
            },
        )

        SERVER_CONFIG.temp_dir.mkdir(parents=True, exist_ok=True)
        temp_path = SERVER_CONFIG.temp_dir / f"{uuid4()}.webm"
        video_bytes = base64.b64decode(base64_data)
        temp_path.write_bytes(video_bytes)

        print(f"[Renderer] Recording saved: {temp_path} ({len(video_bytes) / 1024 / 1024:.2f} MB)")
        return str(temp_path)
    finally:
        if page:
            try:
                await page.close()
            except Exception:
                pass
        _active_firefly_video_records -= 1


def get_active_firefly_video_record_count() -> int:
    return _active_firefly_video_records


async def close_browser() -> None:
    global _playwright, _browser
    if _browser:
        try:
            await _browser.close()
        except Exception:
            pass
        _browser = None
    if _playwright:
        try:
            await _playwright.stop()
        except Exception:
            pass
        _playwright = None

File: services/renderer.js
/**
 * Web2Media Service — Puppeteer Rendering Service
 * 
 * Uses a singleton browser instance for performance.
 * Each recording request gets a new page (tab).
 */

const puppeteer = require('puppeteer');
const path = require('path');
const fs = require('fs');
const { v4: uuidv4 } = require('uuid');
const { SERVER_CONFIG } = require('../config');

let browserInstance = null;
let activeRecordings = 0;

/**
 * Download an image from URL and convert to base64 data URL (server-side).
 * Handles redirects, timeouts, and sends a browser-like User-Agent.
 * 
 * @param {string} url - Image URL to download
 * @param {number} maxRedirects - Max redirect hops
 * @returns {Promise<string>} Base64 data URL
 */
function downloadImageAsDataUrl(url, maxRedirects = 5) {
  return new Promise((resolve, reject) => {
    if (maxRedirects <= 0) return reject(new Error('Too many redirects'));

    const protocol = url.startsWith('https') ? require('https') : require('http');
    const options = {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
        'Accept': 'image/*,*/*',
        'Referer': new URL(url).origin,
      },
      timeout: 15000,
    };

    const req = protocol.get(url, options, (res) => {
      // Follow redirects
      if ([301, 302, 303, 307, 308].includes(res.statusCode) && res.headers.location) {
        const redirectUrl = new URL(res.headers.location, url).href;
        return downloadImageAsDataUrl(redirectUrl, maxRedirects - 1).then(resolve, reject);
      }

      if (res.statusCode !== 200) {
        return reject(new Error(`HTTP ${res.statusCode} from ${url}`));
      }

      const contentType = res.headers['content-type'] || 'image/jpeg';
      const chunks = [];

      res.on('data', (chunk) => chunks.push(chunk));
      res.on('end', () => {
        const buffer = Buffer.concat(chunks);
        if (buffer.length < 100) {
          return reject(new Error('Downloaded file too small — likely not an image'));
        }
        const base64 = buffer.toString('base64');
        const mimeType = contentType.split(';')[0].trim();
        resolve(`data:${mimeType};base64,${base64}`);
      });
      res.on('error', reject);
    });

    req.on('timeout', () => { req.destroy(); reject(new Error('Download timeout (15s)')); });
    req.on('error', reject);
  });
}

/**
 * Get or launch the singleton browser instance
 */
async function getBrowser() {
  if (!browserInstance || !browserInstance.connected) {
    browserInstance = await puppeteer.launch({
      headless: 'new',
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-gpu',
        '--no-first-run',
        '--no-zygote',
        '--disable-extensions',
        '--autoplay-policy=no-user-gesture-required',
      ],
    });

    // Handle unexpected browser close
    browserInstance.on('disconnected', () => {
      browserInstance = null;
    });
  }
  return browserInstance;
}

/**
 * Render a firefly animation video
 * 
 * @param {Object} params - Recording parameters
 * @param {number} params.count - Number of fireflies
 * @param {number} params.size - Firefly size
 * @param {number} params.speed - Firefly speed
 * @param {string} params.colorMode - 'preset' or 'custom'
 * @param {number} params.colorIndex - Color preset index
 * @param {string} params.customColor - Custom hex color
 * @param {string} params.glowLevel - Glow intensity
 * @param {string} params.direction - Movement direction
 * @param {number} params.spread - Spread factor
 * @param {number} params.bgIndex - Background preset index
 * @param {string|null} params.bgUrl - Custom background URL
 * @param {number} params.duration - Recording duration in seconds
 * @param {number} params.fps - Frames per second
 * @param {number} params.width - Video width
 * @param {number} params.height - Video height
 * @param {number} params.bitrate - Video bitrate in bps
 * @returns {Promise<string>} Path to the recorded .webm file
 */
async function renderVideo(params) {
  // Check concurrent limit
  if (activeRecordings >= SERVER_CONFIG.maxConcurrent) {
    throw new Error(`Đã đạt giới hạn ${SERVER_CONFIG.maxConcurrent} video đồng thời. Vui lòng thử lại sau.`);
  }

  activeRecordings++;
  let page = null;

  try {
    const browser = await getBrowser();
    page = await browser.newPage();

    // Set viewport to requested resolution
    await page.setViewport({
      width: params.width,
      height: params.height,
      deviceScaleFactor: 1,
    });

    // Navigate to the render page
    const renderPagePath = path.join(SERVER_CONFIG.publicDir, 'firefly-render.html');
    await page.goto(`file://${renderPagePath}`, {
      waitUntil: 'domcontentloaded',
      timeout: 15000,
    });

    // Wait for page to be ready
    await page.waitForFunction('window.__PAGE_READY === true', { timeout: 10000 });

    // Build config for the firefly engine
    const fireflyConfig = {
      bgIndex: params.bgIndex,
      bgCustom: params.bgUrl || null,
      count: params.count,
      size: params.size,
      speed: params.speed,
      colorMode: params.colorMode,
      colorIndex: params.colorIndex,
      customColor: params.customColor,
      glowLevel: params.glowLevel,
      direction: params.direction,
      spread: params.spread,
    };

    // If bgUrl is provided, download image server-side (bypasses CORS/bot detection)
    if (params.bgUrl) {
      console.log(`[Renderer] Downloading background: ${params.bgUrl}`);
      try {
        const dataUrl = await downloadImageAsDataUrl(params.bgUrl);
        fireflyConfig.bgCustom = dataUrl;
        console.log(`[Renderer] Background loaded (${(dataUrl.length / 1024).toFixed(0)} KB data URL)`);
      } catch (dlErr) {
        console.warn(`[Renderer] Background download failed: ${dlErr.message}. Using preset instead.`);
        // Fallback to preset background — don't fail the whole recording
        fireflyConfig.bgCustom = null;
      }
    }

    // Apply config
    await page.evaluate((cfg) => {
      window.__applyConfig(cfg);
    }, fireflyConfig);

    // Wait a bit for fireflies to initialize and spread
    await new Promise(r => setTimeout(r, 1500));

    // Start recording
    console.log(`[Renderer] Starting recording: ${params.width}x${params.height}, ${params.duration}s, ${params.fps}fps, ${params.bitrate}bps`);

    const base64Data = await page.evaluate(
      (duration, fps, bitrate) => window.__startRecording(duration, fps, bitrate),
      params.duration,
      params.fps,
      params.bitrate
    );

    // Save to temp file
    const tempId = uuidv4();
    const tempPath = path.join(SERVER_CONFIG.tempDir, `${tempId}.webm`);

    // Ensure temp dir exists
    if (!fs.existsSync(SERVER_CONFIG.tempDir)) {
      fs.mkdirSync(SERVER_CONFIG.tempDir, { recursive: true });
    }

    const buffer = Buffer.from(base64Data, 'base64');
    fs.writeFileSync(tempPath, buffer);

    console.log(`[Renderer] Recording saved: ${tempPath} (${(buffer.length / 1024 / 1024).toFixed(2)} MB)`);

    return tempPath;

  } finally {
    // Close the page
    if (page) {
      try { await page.close(); } catch (e) { /* ignore */ }
    }
    activeRecordings--;
  }
}

/**
 * Render HTML content to a PNG buffer using Puppeteer.
 *
 * @param {string} htmlContent - Full HTML string to render
 * @returns {Promise<Buffer>} PNG image as a Buffer
 */
async function renderThumbnail(htmlContent) {
  let page = null;
  try {
    const browser = await getBrowser();
    page = await browser.newPage();

    await page.setViewport({
      width: SERVER_CONFIG.thumbnailViewport.width,
      height: SERVER_CONFIG.thumbnailViewport.height,
      deviceScaleFactor: 1,
    });

    // Load HTML content directly
    await page.setContent(htmlContent, { waitUntil: 'networkidle0' });

    // Extra wait to ensure font rendering is complete
    await new Promise(r => setTimeout(r, SERVER_CONFIG.fontLoadWait));

    // Screenshot only the #thumbnail element
    const thumbnailElement = await page.$('#thumbnail');

    let buffer;
    if (thumbnailElement) {
      buffer = await thumbnailElement.screenshot({ type: 'png' });
    } else {
      // Fallback: screenshot with clip
      buffer = await page.screenshot({
        type: 'png',
        clip: { x: 0, y: 0, ...SERVER_CONFIG.thumbnailViewport },
      });
    }

    // Puppeteer returns a Uint8Array (Buffer in Node.js)
    return buffer;
  } finally {
    if (page) {
      try { await page.close(); } catch (e) { /* ignore */ }
    }
  }
}

/**
 * Get current active recording count
 */
function getActiveCount() {
  return activeRecordings;
}

/**
 * Cleanup: close browser instance
 */
async function closeBrowser() {
  if (browserInstance) {
    try { await browserInstance.close(); } catch (e) { /* ignore */ }
    browserInstance = null;
  }
}

module.exports = {
  renderVideo,
  renderThumbnail,
  getActiveCount,
  closeBrowser,
};
